File: csharp/DataStructures/src/dsa/tree/BinaryTree.cs
namespace dsa.tree;

public class BinaryTree
{
    // static void Main()
    // {
    //     BinaryTree bt = new();
    //     bt.createBinaryTree();
    //     bt.postOrder();
    // }

    // Nó raiz da árvore
    public TreeNode root;

    // Classe interna para representar um nó da árvore
    public class TreeNode
    {
        public TreeNode left;  // Nó filho à esquerda
        public TreeNode right; // Nó filho à direita
        public int data;       // Dado armazenado no nó

        // Construtor do nó
        public TreeNode(int data) => this.data = data;
    }

    // Método de pré-ordem recursivo
    public void preOrder(TreeNode root)
    {
        if (root == null)
            return;

        Console.Write(root.data + " "); // Processa o nó atual
        preOrder(root.left);            // Processa o nó à esquerda
        preOrder(root.right);           // Processa o nó à direita
    }

    // Método de pré-ordem iterativo
    public void preOrder()
    {
        if (root == null)
            return;

        Stack<TreeNode> stack = new();
        stack.Push(root);

        while (stack.Count != 0)
        {
            TreeNode temp = stack.Pop();
            Console.Write(temp.data + " ");
            if (temp.right != null)
                stack.Push(temp.right);

            if (temp.left != null)
                stack.Push(temp.left);
        }
    }

    // Método de ordem simétrica recursivo
    public void inOrder(TreeNode root)
    {
        if (root == null)
            return;

        inOrder(root.left);             // Processa o nó à esquerda
        Console.Write(root.data + " "); // Processa o nó atual
        inOrder(root.right);            // Processa o nó à direita
    }

    // Método de ordem simétrica iterativo
    public void inOrder()
    {
        if (root == null)
            return;

        Stack<TreeNode> stack = new();
        TreeNode temp = root;

        while (stack.Count != 0 || temp != null)
        {
            if (temp != null)
            {
                stack.Push(temp);
                temp = temp.left;
            }
            else
            {
                temp = stack.Pop();
                Console.Write(temp.data + " ");
                temp = temp.right;
            }
        }
    }

    // Método de pós-ordem recursivo
    public void postOrder(TreeNode root)
    {
        if (root == null)
            return;

        postOrder(root.left);           // Processa o nó à esquerda
        postOrder(root.right);          // Processa o nó à direita
        Console.Write(root.data + " "); // Processa o nó atual
    }

    // Método de pós-ordem iterativo
    public void postOrder()
    {
        TreeNode current = root;
        Stack<TreeNode> stack = new();

        while (current != null || stack.Count != 0)
        {
            if (current != null)
            {
                stack.Push(current);
                current = current.left;
            }
            else
            {
                TreeNode temp = stack.Peek().right;
                if (temp == null)
                {
                    temp = stack.Pop();
                    Console.Write(temp.data + " ");
                    while (stack.Count != 0 && temp == stack.Peek().right)
                    {
                        temp = stack.Pop();
                        Console.Write(temp.data + " ");
                    }
                }
                else
                    current = temp;
            }
        }
    }

    // Método de ordem de nível
    public void levelOrder()
    {
        if (root == null)
            return;

        Queue<TreeNode> queue = new();
        queue.Enqueue(root);

        while (queue.Count != 0)
        {
            TreeNode temp = queue.Dequeue();
            Console.Write(temp.data + " ");
            if (temp.left != null)
                queue.Enqueue(temp.left);

            if (temp.right != null)
                queue.Enqueue(temp.right);
        }
    }

    // Método para encontrar o valor máximo na árvore
    public int findMax() => findMax(root);

    // Método recursivo para encontrar o valor máximo
    public int findMax(TreeNode root)
    {
        if (root == null)
            return int.MinValue;

        int result = root.data;
        int left = findMax(root.left);
        int right = findMax(root.right);

        if (left > result)
            result = left;

        if (right > result)
            result = right;

        return result;
    }

    // Método para criar a árvore binária
    public void createBinaryTree()
    {
        TreeNode first = new(1);
        TreeNode second = new(2);
        TreeNode third = new(3);
        TreeNode fourth = new(4);
        TreeNode fifth = new(5);
        TreeNode sixth = new(6);
        TreeNode seventh = new(7);

        root = first;
        first.left = second;
        first.right = third;
        second.left = fourth;
        second.right = fifth;
        third.left = sixth;
        third.right = seventh;
    }
}

File: go/dsa/tree/BinaryTree/BinaryTree.go
package binarytree

import (
    "container/list"
    "fmt"
    "math"
)

// BinaryTree é a estrutura principal da árvore binária
type BinaryTree struct {
    Root *TreeNode // Nó raiz da árvore
}

// NewBinaryTree cria e retorna uma nova instância de BinaryTree
func NewBinaryTree() *BinaryTree {
    return &BinaryTree{}
}

// TreeNode representa um nó da árvore
type TreeNode struct {
    Data  int       // Dado armazenado no nó
    Left  *TreeNode // Nó filho à esquerda
    Right *TreeNode // Nó filho à direita
}

// CreateBinaryTree cria uma árvore binária com valores fixos
func (bt *BinaryTree) CreateBinaryTree() {
    first := &TreeNode{Data: 1}
    second := &TreeNode{Data: 2}
    third := &TreeNode{Data: 3}
    fourth := &TreeNode{Data: 4}
    fifth := &TreeNode{Data: 5}
    sixth := &TreeNode{Data: 6}
    seventh := &TreeNode{Data: 7}

    bt.Root = first
    first.Left = second
    first.Right = third
    second.Left = fourth
    second.Right = fifth
    third.Left = sixth
    third.Right = seventh
}

// PreOrder realiza a travessia pré-ordem de forma recursiva
func (bt *BinaryTree) PreOrder(root *TreeNode) {
    if root == nil {
        return
    }
    fmt.Print(root.Data, " ")
    bt.PreOrder(root.Left)
    bt.PreOrder(root.Right)
}

// PreOrderIterative realiza a travessia pré-ordem de forma iterativa
func (bt *BinaryTree) PreOrderIterative() {
    if bt.Root == nil {
        return
    }

    stack := list.New()
    stack.PushBack(bt.Root)

    for stack.Len() > 0 {
        node := stack.Remove(stack.Back()).(*TreeNode)
        fmt.Print(node.Data, " ")
        if node.Right != nil {
            stack.PushBack(node.Right)
        }
        if node.Left != nil {
            stack.PushBack(node.Left)
        }
    }
}

// InOrder realiza a travessia in-ordem de forma recursiva
func (bt *BinaryTree) InOrder(root *TreeNode) {
    if root == nil {
        return
    }
    bt.InOrder(root.Left)
    fmt.Print(root.Data, " ")
    bt.InOrder(root.Right)
}

// InOrderIterative realiza a travessia in-ordem de forma iterativa
func (bt *BinaryTree) InOrderIterative() {
    stack := list.New()
    current := bt.Root

    for current != nil || stack.Len() > 0 {
        for current != nil {
            stack.PushBack(current)
            current = current.Left
        }
        current = stack.Remove(stack.Back()).(*TreeNode)
        fmt.Print(current.Data, " ")
        current = current.Right
    }
}

// PostOrder realiza a travessia pós-ordem de forma recursiva
func (bt *BinaryTree) PostOrder(root *TreeNode) {
    if root == nil {
        return
    }
    bt.PostOrder(root.Left)
    bt.PostOrder(root.Right)
    fmt.Print(root.Data, " ")
}

// PostOrderIterative realiza a travessia pós-ordem de forma iterativa
func (bt *BinaryTree) PostOrderIterative() {
    stack := list.New()
    current := bt.Root
    var lastVisited *TreeNode

    for current != nil || stack.Len() > 0 {
        if current != nil {
            stack.PushBack(current)
            current = current.Left
        } else {
            peekNode := stack.Back().Value.(*TreeNode)
            if peekNode.Right != nil && lastVisited != peekNode.Right {
                current = peekNode.Right
            } else {
                fmt.Print(peekNode.Data, " ")
                lastVisited = stack.Remove(stack.Back()).(*TreeNode)
            }
        }
    }
}

// LevelOrder realiza a travessia por nível (BFS)
func (bt *BinaryTree) LevelOrder() {
    if bt.Root == nil {
        return
    }

    queue := list.New()
    queue.PushBack(bt.Root)

    for queue.Len() > 0 {
        node := queue.Remove(queue.Front()).(*TreeNode)
        fmt.Print(node.Data, " ")
        if node.Left != nil {
            queue.PushBack(node.Left)
        }
        if node.Right != nil {
            queue.PushBack(node.Right)
        }
    }
}

// FindMax encontra o valor máximo na árvore
func (bt *BinaryTree) FindMax(root *TreeNode) int {
    if root == nil {
        return math.MinInt
    }

    result := root.Data
    left := bt.FindMax(root.Left)
    right := bt.FindMax(root.Right)

    if left > result {
        result = left
    }
    if right > result {
        result = right
    }
    return result
}
